### ISSUE TITLE
Metric names and statistics (cpu_load filter is broken)

### ISSUE THREAD

**Issue by giesselmann** [ClusterCockpit/cc-backend#111]
Hi, 
I was wondering why filtering for cpu_load didn't work, the fix is quite simple, but points to a slightly more complex problem (Maybe mostly missing documentation). Therefore the issue and not only a small PR:

`git diff internal/repository/job.go`
```
@@ -412,7 +412,7 @@ func (r *JobRepository) MarkArchived(
                        stmt = stmt.Set("mem_used_max", stats.Max)
                case "mem_bw":
                        stmt = stmt.Set("mem_bw_avg", stats.Avg)
-               case "load":
+               case "cpu_load":
                        stmt = stmt.Set("load_avg", stats.Avg)
                case "net_bw":
                        stmt = stmt.Set("net_bw_avg", stats.Avg)
```

cpu_load is also missing in `HandleImportFlag` and `InitDB` within `internal/repository/init.go` and should be added there (I think).

I tried to understand the definition and usage of required metrics and metric statistics, but lost the overview:

- In `internal/repository/job.go` flops_any, mem_used etc are hard-coded in a switch-case
- In `web/frontend/src/filters/Stats.svelte` flops, mem_bw,load and mem_used are hard-coded for filtering
- In `pkg/schema/schemas/job-meta.schema.json` cpu_used, mem_used, ... are marked as required, cpu_load is not (should be insted of cpu_used maybe), other metrics have `(required)` in the description.
- In `pkg/schema/schemas/job-data.schema.json`  some metrics are marked as required with a certain level e.g. node.

Not sure if this list is complete!

So far I thought, monitored metrics and metric-names are mostly site-specific, but there seems to be a set of required metrics (and names, since hard-coded), while others have at least intended but not documented usage.
Example is the `filesystems` definition in `job-data.schema.json`, atm, we push plain lustre_open/close/bw into our cc, but there seems to be a more elegant way of handling also multiple filesystems.

**Comment by moebiusband73** [comment]
Hi,

you are right. This is indeed a bit messy at the moment.

The truth should be the json schemas.  We tried to settle on a defined set of metric and related to this consistent naming.  That is already a difficult undertaking. 

Currently the set of required metrics is:
-  cpu_user
- mem_used
- flops_any
- mem_bw
in job meta for the statistics metrics. And
-  cpu_user
- mem_used
- flops_any
- mem_bw
- net_bw
- filesystems
in job data.
There was a lengthy discussion on what cpu metric is best. We use cpu_load 1m for e.g. finding job allocation errors. Others said that cpu_user is much better, or cpu_used, which is all-idle. 
I still think that at our site most people still look at cpu_load.

The other question is how it is implemented in cc-backend. I tried to follow the schema as close as possible, but e.g. the filesystems array is still not implemented in cc-backend. And as you also figured out there are inconsistencies and hardcoded metric names in various locations.

The problem with the filtered metrics is that they they need to be part of the db schema, that should be equivalent to the job-meta json schema. Every metric you need to filter for needs to be also added to the db schema.
The db schema is currently:
- mem_used_max        REAL NOT NULL DEFAULT 0.0,
- flops_any_avg       REAL NOT NULL DEFAULT 0.0,
- mem_bw_avg          REAL NOT NULL DEFAULT 0.0,
-  load_avg            REAL NOT NULL DEFAULT 0.0,
- net_bw_avg          REAL NOT NULL DEFAULT 0.0,
- net_data_vol_total  REAL NOT NULL DEFAULT 0.0,
- file_bw_avg         REAL NOT NULL DEFAULT 0.0,
- file_data_vol_total REAL NOT NULL DEFAULT 0.0);

So there you have load (and not cpu_load as it should be named), but no cpu_user.
Also the file_bw does not match well with the fact that today there are often multiple relevant network filesystems.
The only required metrics, say cc-backend does not work properly without, are flops_any and mem_bw.

So my thoughts on this:
- I would agree to add cpu_load. But do we need then cpu_load AND cpu_user? Probably yes.
- I suggest to drop the file_bw entries in the db schema and in the job meta required metrics. We have to find some other way there. Do you have a suggestion?
- And of course the implementation should be consistent with the schema
- I cannot promise that we implement the filesystems array support before the first release. This is something bigger as currently cc-backend and frontend is only aware of plain metric lists. We have to discuss how to implement it. Maybe treat filesystems metrics as plain lists by prepending a prefix to them. That would probably be the simplest solution.

I currently work on the buggy import flag implementation and will pull those meta tasks integration multiple modules in a standalone package named importer. The initDB handler will also be placed there. As soon as this is merged into master, we can start a branch to fix the issues with hardcoded metrics and filtering and (if it is not too difficult) supporting the filesystems as a array of metric groups.

**Comment by giesselmann** [comment]
I haven't looked into `cpu_user` so far, we use `cpu_load` for instance as indicator, if code was compiled without openmp.
I think through the code `cpu_load` is more common and also what is stored in the job.db. Not sure if renaming a column is possible with the db migration workflow, if it's easy, worth to clean up.

`File_bw` is part of the ui-defaults, we removed it there, since we had no monitoring for it at the beginning. 
I think for a very basic cc setup, cpu_load, flops_any, mem_used and mem_bw could already be sufficient and allow to display polar and roofline plots.

Everything else, net_bw, file_bw, accelerators is add-on imo.

Deriving most of the implementation from the schema would be great, but taking the `MarkArchived()` function as an example, all metrics except mem_used store an avg, mem_used makes more sense with max.

The filesystem array is imo nothing must-have for 1.0, the definition and naming of a minimal metric-set is more important I would say.

I'm happy to help fixing this topic once the db/archive migration is done.

**Comment by moebiusband73** [comment]
I added cpu_load to all locations you mentioned. I also added mem_used_max to import and initDB.
cpu_load was added as required to the schema.
I opened ticket #146  to keep track of the filesystem schema support.
